File: android/app/src/main/java/dev/mskelton/versly/api/VerslyService.kt
package dev.mskelton.versly.api

import androidx.compose.runtime.compositionLocalOf
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Streaming

const val BASE_URL = "https://versly.mskelton.dev/"

data class TranslationInfo(val id: String, val name: String, val lastUpdated: String)

data class SearchResult(
    val book: String,
    val translationId: String,
    val chapter: String,
    val range: List<String>,
    val relevance: Float? = null,
)

data class SearchResponse(val results: List<SearchResult>)

interface VerslyService {
    @GET("api/translations") suspend fun getTranslations(): Response<List<TranslationInfo>>

    @GET("api/download/{translation}")
    @Streaming
    suspend fun downloadTranslation(
        @Path("translation") translation: String
    ): Response<ResponseBody>

    @GET("api/search")
    suspend fun search(
        @Query("q") query: String,
        @Query("translation") translation: String = "ESV",
    ): Response<SearchResponse>
}

val LocalVerslyService = compositionLocalOf<VerslyService> { error("No VerslyService provided") }

File: ios/Versly/VerslyService.swift
import Foundation

class VerslyService: ObservableObject {
    private let baseURL = "https://versly.mskelton.dev/api/"

    func downloadTranslation(_ translation: String) async throws -> Data {
        guard let url = URL(string: "\(baseURL)download/\(translation)") else {
            throw URLError(.badURL)
        }

        let (data, response) = try await URLSession.shared.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
            httpResponse.statusCode == 200
        else {
            throw URLError(.badServerResponse)
        }

        return data
    }
}
